Improve `Au(0)` constructions

hackerbirds opened the pull request:
<!-- Please describe your changes on the following line: -->

I replaced all `Au(0)` constructions to `Au::zero()` (or other utility functions when possible), except in the legacy layout. This ended up changing some line lengths enough that I had to re-run `./mach fmt` which is the cause for the few formatting changes in the commit.

Upsides:
 - Unify all `Au(0)`s down to a single implementation
 - More readable code (imo)

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because no logic has been modified

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson commented:
I think it's fine to do this for `layout_2020`, but I don't think I'd make too many unnecessary changes to legacy layout. Do you mind reverting those parts of this PR? Thanks.

hackerbirds commented:
> I think it's fine to do this for `layout_2020`, but I don't think I'd make too many unnecessary changes to legacy layout. Do you mind reverting those parts of this PR? Thanks.

Sure no problem!

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

hackerbirds commented:
Sorry about fucking up the rebase, it should be good now :sweat:

mrobinson commented:
@hackerbirds Looks good. Please be sure to sign your commit as well. More details at https://github.com/servo/servo/pull/33709/checks?check_run_id=31251106854.

mrobinson commented:
@hackerbirds Thanks. I've updated the PR description and removed the link about longer linking time. I don't see how this can possibly affect linking time.

Loirooriol commented on components/fonts/shaper.rs:
```diff
@@ -107,16 +108,17 @@ impl ShapedGlyphData {
-                    *y_pos -= y_advance;
-                }
+            let offset =
+                if x_offset == Au::zero() && y_offset == Au::zero() && y_advance == Au::zero() {
```
```suggestion
                if x_offset.is_zero() && y_offset.is_zero() && y_advance.is_zero() {
```

hackerbirds replied:
This function is present in `num_traits`'s `Zero` trait, but I'm using `euclid`'s `Zero` trait which doesn't export `.is_zero()`. Should I add `num_traits`/the feature flag in `app_units` instead by modifying `Cargo.toml` or leave it as is?

hackerbirds replied:
Perhaps I should use `style::Zero` instead?

Loirooriol replied:
Ah, yeah, that's annoying, you can leave it as is

mrobinson replied:
Hrm. You should import `num_traits::Zero` here instead of the `euclid` version for this to work properly. `Au` does implement that and it includes `is_zero()`.

hackerbirds replied:
> Hrm. You should import `num_traits::Zero` here instead of the `euclid` version for this to work properly. `Au` does implement that and it includes `is_zero()`.

It does, but `components/fonts` doesn't have the `num-traits` dependency inside its `Cargo.toml`, so I'd need to add it before using that trait IIRC. If that's ok to add a dependency here though I can do that

mrobinson replied:
It's okay to do that. We already depend on `num-traits` so this isn't adding else to the build chain other than perhaps modifying the crate build order (though unlikely).

hackerbirds replied:
Ok, it's done now

Loirooriol commented on components/layout_2020/table/layout.rs:
```diff
@@ -1044,7 +1044,7 @@ impl<'a> TableLayout<'a> {
             self.column_measures[*column_index]
                 .content_sizes
                 .max_content !=
-                Au(0)
+                Au::zero()
```
Ditto

hackerbirds replied:
This works here because style::Zero is used in this module, but let me know what I should do about the other case.

Loirooriol commented on components/layout_2020/table/layout.rs:
```diff
@@ -2862,8 +2862,8 @@ fn get_outer_sizes_for_measurement(
     (
-        outer_size(size.map(|v| get_size_for_axis(v).unwrap_or(Au(0)))),
-        outer_size(min_size.map(|v| get_size_for_axis(v).unwrap_or(Au(0)))),
+        outer_size(size.map(|v| get_size_for_axis(v).unwrap_or(Au::zero()))),
+        outer_size(min_size.map(|v| get_size_for_axis(v).unwrap_or(Au::zero()))),
```
Maybe `unwrap_or_else(Au::zero)` to reduce the number of parentheses? But either is fine.

hackerbirds replied:
I do prefer the `unwrap_or_else` way

Loirooriol commented on components/layout_thread/lib.rs:
```diff
@@ -875,7 +876,8 @@ impl LayoutThread {
 
-                    let origin = Rect::new(Point2D::new(Au(0), Au(0)), root_size).to_layout();
+                    let origin =
+                        Rect::new(Point2D::new(Au::zero(), Au::zero()), root_size).to_layout();
```
I guess this could use `Point2D::default()`

hackerbirds replied:
`Point2D::zero()` or `Point2D::origin()` also seem to work. I think I will go with `Point2D::zero()` because it's more obvious what it does.

hackerbirds commented:
@Loirooriol thank you! I implemented your suggestions now

mrobinson reviewed (changes requested):
Looks good to me,

hackerbirds committed:
Improve `Au(0)` constructions

This replaces `Au(0)` with `Au::zero()` and other utility functions when possible.

Signed-off-by: hackerbirds <120066692+hackerbirds@users.noreply.github.com>

mrobinson reviewed (approved):
Thank you!
